pkg/apis/azure/validation/infrastructure.go

```diff
@@ -68,23 +68,72 @@ func ValidateInfrastructureConfig(infra *apisazure.InfrastructureConfig, nodesCI
 
 	if infra.Identity != nil && (infra.Identity.Name == "" || infra.Identity.ResourceGroup == "") {
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("identity"), infra.Identity, "specifying an identity requires the name of the identity and the resource group which hosts the identity"))
 	}
 
-	if nodes != nil {
-		allErrs = append(allErrs, nodes.ValidateSubset(workerCIDR)...)
+	return allErrs
+}
+
+func validateNatGatewayConfig(natGatewayConfig *apisazure.NatGatewayConfig, zoned bool, hasVmoAlphaAnnotation bool, natGatewayPath *field.Path) field.ErrorList {
+	var allErrs = field.ErrorList{}
+
+	if natGatewayConfig == nil {
+		return nil
+	}
+
+	if !natGatewayConfig.Enabled {
+		if natGatewayConfig.Zone != nil || natGatewayConfig.IdleConnectionTimeoutMinutes != nil || natGatewayConfig.IPAddresses != nil {
+			return append(allErrs, field.Invalid(natGatewayPath, natGatewayConfig, "NatGateway is disabled but additional NatGateway config is passed"))
+		}
+		return nil
+	}
+
+	// NatGateway cannot be offered for Shoot clusters with a primary AvailabilitySet.
+	// The NatGateway is not compatible with the Basic SKU Loadbalancers which are
+	// required to use for Shoot clusters with AvailabilitySet.
+	if !zoned && !hasVmoAlphaAnnotation {
+		return append(allErrs, field.Forbidden(natGatewayPath, "NatGateway is currently only supported for zonal and VMO clusters"))
+	}
+
+	if natGatewayConfig.IdleConnectionTimeoutMinutes != nil && (*natGatewayConfig.IdleConnectionTimeoutMinutes < natGatewayMinTimeoutInMinutes || *natGatewayConfig.IdleConnectionTimeoutMinutes > natGatewayMaxTimeoutInMinutes) {
+		allErrs = append(allErrs, field.Invalid(natGatewayPath.Child("idleConnectionTimeoutMinutes"), *natGatewayConfig.IdleConnectionTimeoutMinutes, "idleConnectionTimeoutMinutes values must range between 4 and 120"))
+	}
+
+	if natGatewayConfig.Zone == nil {
+		if len(natGatewayConfig.IPAddresses) > 0 {
+			allErrs = append(allErrs, field.Invalid(natGatewayPath.Child("zone"), *natGatewayConfig, "Public IPs can only be selected for zonal NatGateways"))
+		}
+		return allErrs
 	}
+	allErrs = append(allErrs, validateNatGatewayIPReference(natGatewayConfig.IPAddresses, *natGatewayConfig.Zone, natGatewayPath.Child("ipAddresses"))...)
 
 	return allErrs
 }
 
+func validateNatGatewayIPReference(references []apisazure.PublicIPReference, zone int32, fldPath *field.Path) field.ErrorList {
+	var allErrs = field.ErrorList{}
+	for i, ref := range references {
+		if ref.Zone != zone {
+			allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("zone"), ref.Zone, fmt.Sprintf("Public IP can't be used as it is not in the same zone as the NatGateway (zone %d)", zone)))
+		}
+		if ref.Name == "" {
+			allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("name"), ref.Name, "Empty name for NatGateway public IP is invalid"))
```

Comment:
Nitpick, but shouldn't the error here be of type `field.Required` ? You are anyway trying to print an empty string for the name here.

Reply:
Two of the tests will also need to be adapted accordingly to change the error assertion.

Reply:
Yes agree they should rather be required then invalid.